community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java

```diff
@@ -1365,18 +1365,26 @@ public boolean unregisterIndexProvider( String name )
     }
 
     /**
-     * Hook that must be called whenever there is an HA mode switch (eg master/slave switch).
+     * Hook that must be called before there is an HA mode switch (eg master/slave switch),
+     * i.e. after state has changed to pending and before state is about to change to the new target state.
      * This must only be called when the database is otherwise inaccessible.
      */
-    public void afterModeSwitch()
+    public void beforeModeSwitch()
```

Comment:
We have recently introduced `ComponentSwitcher` that can switch component modes depending from slave/master/pending state. It looks like that instead of introducing another `before` listener we can reuse same concept here. WDYT?